src/rpc/mining.cpp

```diff
@@ -578,24 +590,38 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
-        LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = boost::get_system_time() + boost::posix_time::minutes(1);
+            checktxtime = boost::get_system_time() + boost::posix_time::seconds(10);
 
             boost::unique_lock<boost::mutex> lock(csBestBlock);
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
-                if (!cvBlockChange.timed_wait(lock, checktxtime))
-                {
-                    // Timeout: Check transactions for update
-                    if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += boost::posix_time::seconds(10);
+                // Before waiting, generate the coinbase for the block following the next
+                // block (since this is cpu-intensive), so that when next block arrives,
+                // we can quickly respond with the template for following block.
+                // Note that the time to create the coinbase tx here does not add to
+                // the 
+                if (!next_coinbase_mtx && IsShieldedMinerAddress(minerAddress)) {
+                    next_coinbase_height = nHeight + 2;
+                    next_coinbase_mtx = CreateCoinbaseTransaction(Params(), CAmount{0}, minerAddress, next_coinbase_height);
                 }
+                // Release the main lock while waiting
+                LEAVE_CRITICAL_SECTION(cs_main);
```

Comment:
Neither `IsShieldedMinerAddress()` nor `CreateCoinbaseTransaction()` require holding the `cs_main` lock, so we could release it first (which makes it even clearer that we are pre-computing the shielded coinbase transaction while waiting), to allow the rest of the node to progress during that 400ms.